File: AutoSubs-App/src-tauri/resources/modules/template_manager.lua
-- ============================================================
-- template_manager.lua — Quản lý template Fusion Title trong Media Pool
-- GetTemplates, ImportTitleFromFile, GetTemplateItem,
-- GetTemplateItemByFolder, CreateTemplateSet
-- ============================================================

local M = {}

-- ===== PATH: Folder chứa .setting templates trên macOS =====
local TITLES_FOLDER_PATH = os.getenv("HOME")
    .. "/Library/Application Support/Blackmagic Design/DaVinci Resolve/Fusion/Templates/Edit/Titles/AutoSubs"

-- ===== GET TEMPLATES =====
-- Lấy danh sách tất cả Text+ template trong Media Pool
function M.GetTemplates(state, helpers)
    local rootFolder = state.mediaPool:GetRootFolder()
    local t = {}
    local hasDefault = false

    helpers.walk_media_pool(rootFolder, function(clip)
        local props = clip:GetClipProperty()
        local clipType = props["Type"]
        if helpers.isMatchingTitle(clipType) then
            local clipName = props["Clip Name"]
            table.insert(t, { label = clipName, value = clipName })
            if clipName == "Default Template" then
                hasDefault = true
            end
        end
    end)

    -- Import Default Template nếu chưa có
    if not hasDefault and tonumber(state.resolve:GetVersion()[1]) >= 19 then
        print("Default template not found. Importing default template...")
        local ok = pcall(function()
            state.mediaPool:ImportFolderFromFile(helpers.join_path(state.assets_path, "subtitle-template.drb"))
            table.insert(t, { label = "Default Template", value = "Default Template" })
        end)
    end

    return t
end

-- ===== IMPORT TITLE FROM FILE =====
-- Import .setting file từ hệ thống vào Media Pool (xóa cache cũ trước)
-- ⭐ Debug chi tiết để bắt lỗi import
function M.ImportTitleFromFile(state, helpers, templateName)
    local filePath = TITLES_FOLDER_PATH .. "/" .. templateName .. ".setting"
    print("[AutoSubs] ═══════════════════════════════════════")
    print("[AutoSubs] 📁 ImportTitleFromFile: '" .. templateName .. "'")
    print("[AutoSubs]   Path: " .. filePath)

    -- ① Kiểm tra file tồn tại + đọc size + header
    local f = io.open(filePath, "r")
    if not f then
        print("[AutoSubs]   ❌ FILE NOT FOUND — io.open() trả về nil")
        print("[AutoSubs]   💡 Kiểm tra: ls \"" .. TITLES_FOLDER_PATH .. "/\"")
        return nil
    end

    -- Đọc file size
    local content = f:read("*a")
    f:close()
    local fileSize = #content
    print("[AutoSubs]   📏 File size: " .. fileSize .. " bytes")

    if fileSize == 0 then
        print("[AutoSubs]   ❌ FILE RỖNG (0 bytes) — file bị corrupt!")
        return nil
    end

    -- Log header (100 ký tự đầu) để kiểm tra format
    local header = content:sub(1, 100):gsub("\n", "\\n"):gsub("\r", "\\r")
    print("[AutoSubs]   📄 Header(100): " .. header)

    -- Kiểm tra format .setting hợp lệ (phải chứa "Composition" hoặc "MediaIn")
    if not content:find("Composition") and not content:find("MediaIn") and not content:find("TextPlus") then
        print("[AutoSubs]   ⚠️ File KHÔNG chứa 'Composition'/'MediaIn'/'TextPlus' — có thể sai format!")
    else
        print("[AutoSubs]   ✅ Format check OK — tìm thấy keyword Fusion hợp lệ")
    end

    -- ② Log trạng thái DaVinci
    local drVersion = state.resolve:GetVersion()
    if drVersion then
        print("[AutoSubs]   🎬 DaVinci version: " .. tostring(drVersion[1]) .. "." .. tostring(drVersion[2]) .. "." .. tostring(drVersion[3]))
    end

    -- ③ Xóa clip cũ cùng tên (cache DaVinci)
    local rootFolder = state.mediaPool:GetRootFolder()
    local oldClips = {}
    helpers.walk_media_pool(rootFolder, function(clip)
        local props = clip:GetClipProperty()
        if props["Clip Name"] == templateName then
            table.insert(oldClips, clip)
        end
    end)
    if #oldClips > 0 then
        print("[AutoSubs]   🗑 Xóa " .. #oldClips .. " clip cũ trùng tên '" .. templateName .. "'...")
        local delOK = state.mediaPool:DeleteClips(oldClips)
        print("[AutoSubs]   🗑 DeleteClips result: " .. tostring(delOK))
    else
        print("[AutoSubs]   (Không có clip cũ trùng tên)")
    end

    -- ④ Set current folder = root trước khi import
    local setFolderOK = state.mediaPool:SetCurrentFolder(rootFolder)
    print("[AutoSubs]   📂 SetCurrentFolder(root): " .. tostring(setFolderOK))
    print("[AutoSubs]   📂 Current folder: " .. tostring(rootFolder:GetName()))

    -- ⑤ Import .setting file
    print("[AutoSubs]   🔄 Gọi mediaPool:ImportMedia({'" .. filePath .. "'})...")
    local importOK, imported = pcall(function()
        return state.mediaPool:ImportMedia({ filePath })
    end)

    -- Log kết quả chi tiết
    if not importOK then
        print("[AutoSubs]   ❌ ImportMedia EXCEPTION: " .. tostring(imported))
        return nil
    end

    if imported == nil then
        print("[AutoSubs]   ❌ ImportMedia trả về nil (không rõ lý do)")
        print("[AutoSubs]   💡 Có thể: file format không compatible với DaVinci version này")
        print("[AutoSubs]   💡 Thử: mở file bằng Fusion → File → Save As → lưu lại")
        return nil
    end

    if type(imported) ~= "table" then
        print("[AutoSubs]   ❌ ImportMedia trả về type='" .. type(imported) .. "' thay vì table")
        print("[AutoSubs]   Giá trị: " .. tostring(imported))
        return nil
    end

    if #imported == 0 then
        print("[AutoSubs]   ❌ ImportMedia trả về table RỖNG (#imported == 0)")
        print("[AutoSubs]   💡 DaVinci nhận file nhưng từ chối import — có thể:")
        print("[AutoSubs]   💡   1. File .setting tạo từ DaVinci version khác (incompatible)")
        print("[AutoSubs]   💡   2. File bị corrupt hoặc thiếu node Composition")
        print("[AutoSubs]   💡   3. MediaPool đang bị lock bởi process khác")
        return nil
    end

    -- ⑥ Thành công! Log clip properties
    local item = imported[1]
    local props = item:GetClipProperty()
    print("[AutoSubs]   ✅ IMPORT THÀNH CÔNG!")
    print("[AutoSubs]   ✅ Clip Name: " .. tostring(props["Clip Name"]))
    print("[AutoSubs]   ✅ Type: " .. tostring(props["Type"]))
    print("[AutoSubs]   ✅ FPS: " .. tostring(props["FPS"]))
    print("[AutoSubs] ═══════════════════════════════════════")
    return item
end

-- ===== GET TEMPLATE ITEM =====
-- Tìm template theo tên clip trong Media Pool (đệ quy)
function M.GetTemplateItem(helpers, folder, templateName)
    local found = nil
    helpers.walk_media_pool(folder, function(clip)
        local props = clip:GetClipProperty()
        if props["Clip Name"] == templateName then
            found = clip
            return true
        end
    end)
    return found
end

-- ===== GET TEMPLATE ITEM BY FOLDER =====
-- Tìm template bằng tên folder → lấy Fusion Title đầu tiên bên trong
function M.GetTemplateItemByFolder(helpers, rootFolder, templateName)
    -- Thử clip name trước
    local found = M.GetTemplateItem(helpers, rootFolder, templateName)
    if found then return found end

    -- Tìm folder tên trùng
    local function findFolderByName(parent, name)
        for _, subfolder in ipairs(parent:GetSubFolderList()) do
            if subfolder:GetName() == name then
                return subfolder
            end
            local deeper = findFolderByName(subfolder, name)
            if deeper then return deeper end
        end
        return nil
    end

    local targetFolder = findFolderByName(rootFolder, templateName)
    if not targetFolder then
        return nil
    end

    -- Scan folder cho Fusion Title
    local firstFusionTitle = nil
    local firstAnyClip = nil
    helpers.walk_media_pool(targetFolder, function(clip)
        local props = clip:GetClipProperty()
        local clipType = props["Type"] or "?"
        if helpers.isMatchingTitle(clipType) and not firstFusionTitle then
            firstFusionTitle = clip
        end
        if not firstAnyClip then
            firstAnyClip = clip
        end
    end)

    return firstFusionTitle or firstAnyClip
end

-- ===== CREATE TEMPLATE SET =====
-- Tạo nhiều template folder (mỗi cái chứa 1 copy Default Template)
function M.CreateTemplateSet(state, helpers, templateNames)
    print("[AutoSubs] Creating template set...")
    local rootFolder = state.mediaPool:GetRootFolder()
    local currentFolder = state.mediaPool:GetCurrentFolder()

    -- Đảm bảo Default Template tồn tại
    local defaultTpl = M.GetTemplateItem(helpers, rootFolder, "Default Template")
    if not defaultTpl then
        pcall(function()
            state.mediaPool:SetCurrentFolder(rootFolder)
            state.mediaPool:ImportFolderFromFile(helpers.join_path(state.assets_path, "subtitle-template.drb"))
        end)
        defaultTpl = M.GetTemplateItem(helpers, rootFolder, "Default Template")
    end

    if not defaultTpl then
        return { error = true, message = "Cannot find or import Default Template" }
    end

    local results = {}
    local drbPath = helpers.join_path(state.assets_path, "subtitle-template.drb")

    for _, name in ipairs(templateNames) do
        local existing = M.GetTemplateItemByFolder(helpers, rootFolder, name)
        if existing then
            table.insert(results, { name = name, status = "exists" })
        else
            local subfolder = state.mediaPool:AddSubFolder(rootFolder, name)
            if subfolder then
                state.mediaPool:SetCurrentFolder(subfolder)
                local ok = pcall(function()
                    state.mediaPool:ImportFolderFromFile(drbPath)
                end)
                if ok then
                    table.insert(results, { name = name, status = "created" })
                else
                    table.insert(results, { name = name, status = "error", message = "Import failed" })
                end
            else
                table.insert(results, { name = name, status = "error", message = "Folder creation failed" })
            end
        end
    end

    if currentFolder then state.mediaPool:SetCurrentFolder(currentFolder) end
    return { success = true, results = results }
end

-- ===== APPLY CUSTOM TEMPLATE TO TIMELINE ITEM =====
-- Strategy (theo ChatGPT analysis):
--   ImportFusionComp() CHỈ hỗ trợ file .comp (Fusion Composition)
--   KHÔNG hỗ trợ file .setting (MacroOperator template)
--
-- Flow:
--   1. Tìm file .comp trước (đã convert từ .setting bằng convert_settings_to_comp.lua)
--   2. Nếu có .comp → dùng ImportFusionComp() (đường chính, đúng format)
--   3. Nếu không có .comp → báo fail, dùng programmatic style ở caller
--
-- Return: true nếu apply thành công, false nếu fail
function M.ApplySettingToTimelineItem(state, timelineItem, templateName)
    print("[AutoSubs] 🎨 ApplySettingToTimelineItem: '" .. templateName .. "'")

    -- ═══ TÌM FILE .comp (ưu tiên — format đúng cho ImportFusionComp) ═══
    local compPath = TITLES_FOLDER_PATH .. "/" .. templateName .. ".comp"
    local settingPath = TITLES_FOLDER_PATH .. "/" .. templateName .. ".setting"

    -- Kiểm tra .comp trước
    local compFile = io.open(compPath, "r")
    if compFile then
        local compContent = compFile:read("*a")
        compFile:close()
        print("[AutoSubs]   📄 Tìm thấy file .comp: " .. #compContent .. " bytes")

        -- Debug: comp count TRƯỚC import
        local countBefore = timelineItem:GetFusionCompCount()
        print("[AutoSubs]   📊 Comp count TRƯỚC import: " .. tostring(countBefore))

        -- Dùng ImportFusionComp với file .comp (format đúng)
        print("[AutoSubs]   🔧 ImportFusionComp (.comp)...")
        local ok, err = pcall(function()
            local result = timelineItem:ImportFusionComp(compPath)
            print("[AutoSubs]   📋 Return type: " .. type(result) .. " = " .. tostring(result))
            if result then
                return true
            end
            return false
        end)

        -- Debug: comp count SAU import
        local countAfter = timelineItem:GetFusionCompCount()
        print("[AutoSubs]   📊 Comp count SAU import: " .. tostring(countAfter))

        if ok and err then
            if countAfter > countBefore then
                print("[AutoSubs]   ✅ Comp count TĂNG " .. countBefore .. " → " .. countAfter .. " (THÀNH CÔNG!)")
            else
                print("[AutoSubs]   ⚠ Comp count KHÔNG TĂNG — có thể .comp cũng không đúng format")
            end

            -- Debug: dump tool list của comp cuối cùng
            local comp = timelineItem:GetFusionCompByIndex(countAfter)
            if comp then
                local tools = comp:GetToolList(false) or {}
                local toolCount = 0
                for _ in pairs(tools) do toolCount = toolCount + 1 end
                print("[AutoSubs]   🔧 Tools trong comp (index " .. countAfter .. "): " .. toolCount)
                for _, tool in pairs(tools) do
                    local attrs = tool:GetAttrs() or {}
                    print("[AutoSubs]     → " .. tostring(attrs.TOOLS_Name) .. " [" .. tostring(attrs.TOOLS_RegID) .. "]")
                end
            end

            return true
        else
            print("[AutoSubs]   ❌ ImportFusionComp(.comp) failed: " .. tostring(err))
        end
    else
        print("[AutoSubs]   ⚠ Không có file .comp: " .. compPath)
    end

    -- ═══ KHÔNG CÒN THỬ .setting NỮA ═══
    -- (ImportFusionComp KHÔNG hỗ trợ .setting MacroOperator format)
    -- Kiểm tra .setting tồn tại để thông báo cho user
    local settingFile = io.open(settingPath, "r")
    if settingFile then
        settingFile:close()
        print("[AutoSubs]   ⚠ File .setting TỒN TẠI nhưng ImportFusionComp không hỗ trợ MacroOperator format")
        print("[AutoSubs]   💡 Cần convert .setting → .comp bằng script: convert_settings_to_comp.lua")
        print("[AutoSubs]   💡 Chạy script trong DaVinci Fusion Console để tạo file .comp")
    else
        print("[AutoSubs]   ❌ Không có cả .comp lẫn .setting cho '" .. templateName .. "'")
    end

    print("[AutoSubs]   ❌ APPLY FAIL → sẽ dùng programmatic style")
    return false
end

-- ===== GET SETTING FILE PATH =====
-- Trả về path đầy đủ tới file .comp hoặc .setting nếu tồn tại
-- Ưu tiên .comp trước
function M.GetSettingFilePath(templateName)
    -- Thử .comp trước
    local compPath = TITLES_FOLDER_PATH .. "/" .. templateName .. ".comp"
    local f1 = io.open(compPath, "r")
    if f1 then
        f1:close()
        return compPath
    end
    -- Fallback .setting
    local settingPath = TITLES_FOLDER_PATH .. "/" .. templateName .. ".setting"
    local f2 = io.open(settingPath, "r")
    if f2 then
        f2:close()
        return settingPath
    end
    return nil
end

return M

File: AutoSubs-App/src-tauri/resources/AutoSubs/template_manager.py
# ============================================================
# template_manager.py — Quản lý templates trong Media Pool
# Bao gồm: tìm template, import template, tạo template set
# ============================================================

import os
from . import state
from .helpers import join_path

# ===== Danh sách các chuỗi nhận diện Fusion Title theo ngôn ngữ =====
# DaVinci Resolve hiển thị Type khác nhau tuỳ ngôn ngữ hệ thống
TITLE_STRINGS = {
    "Título – Fusion",      # Spanish
    "Título Fusion",         # Portuguese
    "Generator",             # English (older versions)
    "Fusion Title",          # English
    "Titre Fusion",          # French
    "Титры на стр. Fusion",  # Russian
    "Fusion Titel",          # German
    "Titolo Fusion",         # Italian
    "Fusionタイトル",          # Japanese
    "Fusion标题",             # Chinese
    "퓨전 타이틀",              # Korean
    "Tiêu đề Fusion",       # Vietnamese
    "Fusion Titles",         # Thai
}

# ===== Đường dẫn folder chứa .setting templates trên macOS =====
# DaVinci Resolve đọc từ thư mục này để hiển thị trong Effects Library → Titles
TITLES_FOLDER_PATH = os.path.join(
    os.path.expanduser("~"),
    "Library/Application Support/Blackmagic Design/DaVinci Resolve/"
    "Fusion/Templates/Edit/Titles/AutoSubs"
)


def is_matching_title(title):
    """Kiểm tra xem Type của clip có phải là Fusion Title không"""
    return title in TITLE_STRINGS


def walk_media_pool(folder, on_clip):
    """
    Duyệt đệ quy toàn bộ Media Pool (folder + subfolder)
    Gọi on_clip(clip) cho mỗi clip tìm được
    Nếu on_clip trả về True → dừng duyệt sớm (early exit)
    """
    # Duyệt subfolder trước (đệ quy)
    subfolders = folder.GetSubFolderList()
    if subfolders:
        for subfolder in subfolders:
            stop = walk_media_pool(subfolder, on_clip)
            if stop:
                return True

    # Duyệt tất cả clip trong folder hiện tại
    clips = folder.GetClipList()
    if clips:
        for clip in clips:
            stop = on_clip(clip)
            if stop:
                return True

    return False


def get_templates():
    """
    Lấy danh sách tất cả Text+ templates trong Media Pool
    Trả về: [{"label": "Template Name", "value": "Template Name"}, ...]
    """
    root_folder = state.media_pool.GetRootFolder()
    templates = []
    has_default = False

    def check_clip(clip):
        nonlocal has_default
        props = clip.GetClipProperty()
        clip_type = props.get("Type", "")
        if is_matching_title(clip_type):
            clip_name = props.get("Clip Name", "")
            templates.append({"label": clip_name, "value": clip_name})
            if clip_name == "Default Template":
                has_default = True
        return False  # tiếp tục duyệt

    walk_media_pool(root_folder, check_clip)

    # Nếu chưa có Default Template → import từ file .drb
    if not has_default:
        try:
            version = state.resolve.GetVersion()
            if version and int(version[0]) >= 19:
                print("[AutoSubs] Default template not found. Importing...")
                drb_path = join_path(state.assets_path, "subtitle-template.drb")
                state.media_pool.ImportFolderFromFile(drb_path)
                templates.append({"label": "Default Template", "value": "Default Template"})
        except Exception as e:
            print(f"[AutoSubs] Failed to import default template: {e}")

    return templates


def get_template_item(folder, template_name):
    """
    Tìm template theo tên clip trong Media Pool
    Duyệt toàn bộ folder + subfolder
    Trả về MediaPoolItem hoặc None
    """
    found = [None]  # dùng list để có thể gán trong closure

    def check_clip(clip):
        props = clip.GetClipProperty()
        if props.get("Clip Name") == template_name:
            found[0] = clip
            return True  # dừng duyệt
        return False

    walk_media_pool(folder, check_clip)
    return found[0]


def get_template_item_by_folder(root_folder, template_name):
    """
    Tìm template bằng TÊN FOLDER trong Media Pool
    Ưu tiên 1: tìm clip có tên trùng (nhanh nhất)
    Ưu tiên 2: tìm folder có tên trùng → lấy Fusion Title đầu tiên bên trong
    """
    # Bước 1: thử tìm bằng tên clip trước
    found = get_template_item(root_folder, template_name)
    if found:
        return found

    # Bước 2: tìm folder có tên khớp
    def find_folder_by_name(parent, name):
        subfolders = parent.GetSubFolderList()
        if subfolders:
            for subfolder in subfolders:
                if subfolder.GetName() == name:
                    return subfolder
                deeper = find_folder_by_name(subfolder, name)
                if deeper:
                    return deeper
        return None

    target_folder = find_folder_by_name(root_folder, template_name)
    if not target_folder:
        print(f"[AutoSubs] Folder '{template_name}' not found in Media Pool")
        return None

    # Quét folder tìm Fusion Title đầu tiên
    print(f"[AutoSubs] Scanning folder '{template_name}' for usable clip...")
    first_fusion_title = [None]
    first_any_clip = [None]

    def check_clip(clip):
        props = clip.GetClipProperty()
        clip_type = props.get("Type", "?")
        clip_name = props.get("Clip Name", "?")
        print(f"[AutoSubs]   Found clip: '{clip_name}' (Type='{clip_type}')")

        # Ưu tiên 1: Fusion Title
        if is_matching_title(clip_type) and first_fusion_title[0] is None:
            first_fusion_title[0] = clip

        # Ưu tiên 2: bất kỳ clip nào (fallback)
        if first_any_clip[0] is None:
            first_any_clip[0] = clip

        return False  # không dừng sớm — in log hết

    walk_media_pool(target_folder, check_clip)

    # Chọn clip tốt nhất
    title_item = first_fusion_title[0] or first_any_clip[0]

    if first_fusion_title[0]:
        props = first_fusion_title[0].GetClipProperty()
        print(f"[AutoSubs] Using Fusion Title: '{props.get('Clip Name', '?')}' "
              f"for template '{template_name}'")
    elif first_any_clip[0]:
        props = first_any_clip[0].GetClipProperty()
        print(f"[AutoSubs] WARNING: No Fusion Title found — using fallback clip: "
              f"'{props.get('Clip Name', '?')}' (Type='{props.get('Type', '?')}')")
    else:
        print(f"[AutoSubs] ERROR: Folder '{template_name}' is completely empty!")

    return title_item


def import_title_from_file(template_name):
    """
    Import file .setting từ folder hệ thống vào Media Pool
    File .setting nằm ở ~/Library/.../Titles/AutoSubs/
    
    QUAN TRỌNG: Xoá clip cũ cùng tên trước → import fresh
    (DaVinci cache clip theo filename, nên phải xoá trước)
    """
    file_path = os.path.join(TITLES_FOLDER_PATH, f"{template_name}.setting")
    print(f"[AutoSubs] Trying to import title from: {file_path}")

    # Kiểm tra file có tồn tại không
    if not os.path.exists(file_path):
        print(f"[AutoSubs] ⚠ File not found: {file_path}")
        return None

    # ===== XOÁ CLIP CŨ CÙNG TÊN TRONG MEDIA POOL =====
    root_folder = state.media_pool.GetRootFolder()
    old_clips = []

    def find_old(clip):
        props = clip.GetClipProperty()
        if props.get("Clip Name") == template_name:
            old_clips.append(clip)
        return False

    walk_media_pool(root_folder, find_old)

    if old_clips:
        print(f"[AutoSubs] 🗑 Deleting {len(old_clips)} old cached clip(s) "
              f"named '{template_name}' from Media Pool...")
        try:
            delete_ok = state.media_pool.DeleteClips(old_clips)
            if delete_ok:
                print("[AutoSubs] ✅ Old clips deleted successfully")
            else:
                print("[AutoSubs] ⚠ DeleteClips returned false — may still be cached")
        except Exception as e:
            print(f"[AutoSubs] ⚠ DeleteClips error: {e}")

    # Import fresh từ file .setting mới nhất
    state.media_pool.SetCurrentFolder(root_folder)
    try:
        imported = state.media_pool.ImportMedia([file_path])
    except Exception as e:
        print(f"[AutoSubs] ❌ ImportMedia failed: {e}")
        return None

    if not imported or len(imported) == 0:
        print(f"[AutoSubs] ❌ ImportMedia failed for: {file_path}")
        return None

    # Trả về item đầu tiên import được
    item = imported[0]
    props = item.GetClipProperty()
    print(f"[AutoSubs] ✅ Imported FRESH title: '{props.get('Clip Name', '?')}' from {file_path}")
    return item


def create_template_set(template_names):
    """
    Tạo nhiều template folder trong Media Pool
    Mỗi folder chứa 1 bản copy của Default Template (import từ .drb)
    User sau đó vào DaVinci Resolve để customize mỗi template riêng
    """
    print("[AutoSubs] Creating template set...")
    root_folder = state.media_pool.GetRootFolder()
    current_folder = state.media_pool.GetCurrentFolder()

    # Đảm bảo Default Template tồn tại trước
    default_tpl = get_template_item(root_folder, "Default Template")
    if not default_tpl:
        print("[AutoSubs] Importing Default Template from .drb...")
        try:
            state.media_pool.SetCurrentFolder(root_folder)
            drb_path = join_path(state.assets_path, "subtitle-template.drb")
            state.media_pool.ImportFolderFromFile(drb_path)
        except Exception:
            pass
        default_tpl = get_template_item(root_folder, "Default Template")

    if not default_tpl:
        return {"error": True, "message": "Cannot find or import Default Template"}

    results = []
    drb_path = join_path(state.assets_path, "subtitle-template.drb")

    for name in template_names:
        # Kiểm tra đã tồn tại chưa
        existing = get_template_item_by_folder(root_folder, name)
        if existing:
            print(f"[AutoSubs] Template '{name}' already exists — skipping")
            results.append({"name": name, "status": "exists"})
        else:
            # Tạo subfolder mới
            print(f"[AutoSubs] Creating folder '{name}'...")
            subfolder = state.media_pool.AddSubFolder(root_folder, name)
            if subfolder:
                state.media_pool.SetCurrentFolder(subfolder)
                try:
                    state.media_pool.ImportFolderFromFile(drb_path)
                    imported = get_template_item_by_folder(root_folder, name)
                    if imported:
                        print(f"[AutoSubs] ✅ Template '{name}' created successfully!")
                        results.append({"name": name, "status": "created"})
                    else:
                        print(f"[AutoSubs] ⚠ Folder created but no template for '{name}'")
                        results.append({"name": name, "status": "error", "message": "Import failed"})
                except Exception:
                    print(f"[AutoSubs] ❌ Failed to import .drb into folder '{name}'")
                    results.append({"name": name, "status": "error", "message": "Import failed"})
            else:
                print(f"[AutoSubs] ❌ Could not create folder '{name}'")
                results.append({"name": name, "status": "error", "message": "Folder creation failed"})

    # Quay lại folder gốc
    if current_folder:
        state.media_pool.SetCurrentFolder(current_folder)

    print("[AutoSubs] Template set creation done.")
    return {"success": True, "results": results}
